src/CodeGen_Hexagon.cpp

```diff
@@ -1319,6 +1319,123 @@ void CodeGen_Hexagon::visit(const Cast *op) {
+        lane = ConstantInt::get(i32_t, i);
+        any_true = builder->CreateOr(any_true, builder->CreateIsNotNull(builder->CreateExtractElement(vpred, lane)));
+    }
+
+    BasicBlock *true_bb = BasicBlock::Create(*context, "true_bb", function);
+    BasicBlock *false_bb = BasicBlock::Create(*context, "false_bb", function);
+    BasicBlock *after_bb = BasicBlock::Create(*context, "after_bb", function);
+    builder->CreateCondBr(any_true, true_bb, false_bb);
+
+    builder->SetInsertPoint(true_bb);
+    Value *true_value = codegen(load_expr);
+    builder->CreateBr(after_bb);
+    // The true value might have jumped to a new block (e.g. if there was a
+    // nested select). To generate a correct PHI node, grab the current
+    // block.
+    BasicBlock *true_pred = builder->GetInsertBlock();
+
+    builder->SetInsertPoint(false_bb);
+    Value *false_value = UndefValue::get(llvm_type_of(load_expr.type()));
+    builder->CreateBr(after_bb);
+    BasicBlock *false_pred = builder->GetInsertBlock();
+
+    builder->SetInsertPoint(after_bb);
+    PHINode *phi = builder->CreatePHI(true_value->getType(), 2);
+    phi->addIncoming(true_value, true_pred);
+    phi->addIncoming(false_value, false_pred);
+
+    value = phi;
+}
+
+void CodeGen_Hexagon::codegen_predicated_vector_store(const Call *op) {
+    const Call *store_addr = op->args[0].as<Call>();
+    internal_assert(store_addr && (store_addr->is_intrinsic(Call::address_of)))
+        << "The first argument to predicated_store must be call to address_of of the store\n";
+
+    const Broadcast *broadcast = store_addr->args[0].as<Broadcast>();
+    const Load *load = broadcast ? broadcast->value.as<Load>() : store_addr->args[0].as<Load>();
+    internal_assert(load) << "The sole argument to address_of must be a load or broadcast of load\n";
+
+    // Even on 32-bit systems, Handles are treated as 64-bit in
+    // memory, so convert stores of handles to stores of uint64_ts.
+    if (op->args[2].type().is_handle()) {
+        Expr v = reinterpret(UInt(64, op->args[2].type().lanes()), op->args[2]);
+        Expr dest = Call::make(Handle().with_lanes(v.type().lanes()), Call::address_of,
+                               {Load::make(v.type(), load->name, load->index, load->image, load->param)},
+                               Call::Intrinsic);
+        Expr expr = Call::make(v.type(), Call::predicated_store,
+                               {dest, op->args[1], v},
+                               Call::Intrinsic);
+        codegen(expr);
+        return;
+    }
+
+    Type value_type = op->args[2].type().element_of();
+
+    // We need to scalarize the predicated store since masked store/load on
+    // hexagon is not handled by the LLVM
+    debug(4) << "Scalarize predicated vector store on hexagon\n";
+    Value *vpred = codegen(op->args[1]);
+    Value *vval = codegen(op->args[2]);
+    Expr index = broadcast ? Broadcast::make(load->index, broadcast->lanes) : load->index;
```

Comment:
Rather than make the index a vector if it is a scalar here, you could instead leave the index as is, and below, conditionally use CreateExtractElement instead. This might generate cleaner code when the index is a scalar.